misc-utils/lsfd-sock-xinfo.c

```diff
@@ -52,41 +55,111 @@ static void load_xinfo_from_proc_udp6(ino_t netns_inode);
 static void load_xinfo_from_proc_udplite6(ino_t netns_inode);
 static void load_xinfo_from_proc_raw6(ino_t netns_inode);
 static void load_xinfo_from_proc_netlink(ino_t netns_inode);
+static void load_xinfo_from_proc_packet(ino_t netns_inode);
 
 static int self_netns_fd = -1;
 static struct stat self_netns_sb;
 
 static void *xinfo_tree;	/* for tsearch/tfind */
 static void *netns_tree;
 
+struct iface {
+	unsigned int index;
+	char name[IF_NAMESIZE];
+};
+
+static const char *get_iface_name(ino_t netns, unsigned int iface_index);
+
+struct netns {
+	ino_t inode;
+	size_t n_ifaces;
+	struct iface *ifaces;
+};
+
 static int netns_compare(const void *a, const void *b)
 {
-	if (*(ino_t *)a < *(ino_t *)b)
-		return -1;
-	else if (*(ino_t *)a > *(ino_t *)b)
-		return 1;
-	else
-		return 0;
+	const struct netns *netns_a = a;
+	const struct netns *netns_b = b;
+
+	return netns_a->inode - netns_b->inode;
+}
+
+static void netns_free(void *netns)
+{
+	struct netns *nsobj = netns;
+
+	if (nsobj->n_ifaces > 0)
+		free(nsobj->ifaces);
+	free(netns);
+}
+
+/*
+ * iface index -> iface name mappings
+ */
+static void load_ifaces_from_getifaddrs(struct netns *nsobj)
+{
+	struct ifaddrs *ifa_list;
+	struct ifaddrs *ifa;
+	const size_t buflen = sizeof(nsobj->ifaces[0].name);
+	size_t i, count = 0;
+
+	if (getifaddrs(&ifa_list) < 0)
+		return;
+
+	for (ifa = ifa_list; ifa != NULL; ifa = ifa->ifa_next)
+		count++;
+
+	nsobj->n_ifaces = count;
+	nsobj->ifaces = xmalloc(count * sizeof(struct iface));
+
+	for (ifa = ifa_list, i = 0; ifa != NULL; ifa = ifa->ifa_next, i++) {
+		unsigned int if_index = if_nametoindex(ifa->ifa_name);
+
+		nsobj->ifaces[i].index = if_index;
+		strncpy(nsobj->ifaces[i].name, ifa->ifa_name, buflen - 1);
+		nsobj->ifaces[i].name[buflen - 1] = '\0';
+	}
+
+	freeifaddrs(ifa_list);
+
+	return;
+}
+
+static const char *get_iface_name(ino_t netns, unsigned int iface_index)
+{
+	struct netns **nsobj = tfind(&netns, &netns_tree, netns_compare);
+	if (! (nsobj && (*nsobj)->n_ifaces))
```

Comment:
if (!nsobj) is clearer and count == 0 will be handled by the loop below

Reply:
Done.